File: StampLetterFlutter/lib/models/letter.dart
import 'package:flutter/material.dart';
import 'stamp.dart';

enum LetterPaperStyle { kraft, watercolor, check, floral, white }

enum StampPosition { topRight, topLeft, bottomRight, bottomLeft }

enum LetterFont { handwriting, serif, sans }

extension LetterPaperStyleX on LetterPaperStyle {
  String get label {
    switch (this) {
      case LetterPaperStyle.kraft:
        return '크라프트';
      case LetterPaperStyle.watercolor:
        return '수채화';
      case LetterPaperStyle.check:
        return '체크';
      case LetterPaperStyle.floral:
        return '플로럴';
      case LetterPaperStyle.white:
        return '화이트';
    }
  }
}

extension StampPositionX on StampPosition {
  String get label {
    switch (this) {
      case StampPosition.topRight:
        return '상단 오른쪽';
      case StampPosition.topLeft:
        return '상단 왼쪽';
      case StampPosition.bottomRight:
        return '하단 오른쪽';
      case StampPosition.bottomLeft:
        return '하단 왼쪽';
    }
  }
}

extension LetterFontX on LetterFont {
  String get label {
    switch (this) {
      case LetterFont.handwriting:
        return '손글씨';
      case LetterFont.serif:
        return '명조';
      case LetterFont.sans:
        return '고딕';
    }
  }

  TextStyle get textStyle {
    switch (this) {
      case LetterFont.handwriting:
        return const TextStyle(fontFamily: 'DancingScript', fontSize: 16);
      case LetterFont.serif:
        return const TextStyle(fontFamily: 'Merriweather', fontSize: 16);
      case LetterFont.sans:
        return const TextStyle(fontFamily: 'NotoSans', fontSize: 16);
    }
  }
}

class Letter {
  final String id;
  final Stamp stamp;
  final LetterPaperStyle paperStyle;
  final StampPosition stampPosition;
  final String content;
  final LetterFont font;
  final Color textColor;
  final String senderName;
  final String receiverName;
  final DateTime createdAt;

  Letter({
    required this.id,
    required this.stamp,
    required this.paperStyle,
    required this.stampPosition,
    required this.content,
    required this.font,
    required this.textColor,
    required this.senderName,
    required this.receiverName,
    required this.createdAt,
  });

  factory Letter.fromJson(Map<String, dynamic> json) {
    return Letter(
      id: json['id'] as String,
      stamp: Stamp.fromJson(json['stamp'] as Map<String, dynamic>),
      paperStyle: LetterPaperStyle.values.firstWhere(
        (style) => style.name == json['paperStyle'] as String,
        orElse: () => LetterPaperStyle.white,
      ),
      stampPosition: StampPosition.values.firstWhere(
        (position) => position.name == json['stampPosition'] as String,
        orElse: () => StampPosition.topRight,
      ),
      content: json['content'] as String,
      font: LetterFont.values.firstWhere(
        (font) => font.name == json['font'] as String,
        orElse: () => LetterFont.handwriting,
      ),
      textColor: Color(int.parse(json['textColor'] as String)),
      senderName: json['senderName'] as String,
      receiverName: json['receiverName'] as String,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'stamp': stamp.toJson(),
      'paperStyle': paperStyle.name,
      'stampPosition': stampPosition.name,
      'content': content,
      'font': font.name,
      'textColor': textColor.value.toString(),
      'senderName': senderName,
      'receiverName': receiverName,
      'createdAt': createdAt.toIso8601String(),
    };
  }
}

File: Models/Letter.swift
import SwiftUI
import UIKit

enum LetterPaperStyle: String, CaseIterable, Codable, Identifiable {
    case kraft
    case watercolor
    case check
    case floral
    case white

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .kraft: return "크라프트"
        case .watercolor: return "수채화"
        case .check: return "체크"
        case .floral: return "플로럴"
        case .white: return "화이트"
        }
    }

    var previewColor: Color {
        switch self {
        case .kraft: return Color(red: 0.83, green: 0.66, blue: 0.42)
        case .watercolor: return Color(red: 0.93, green: 0.89, blue: 0.98)
        case .check: return Color(.systemBackground)
        case .floral: return Color(red: 0.98, green: 0.95, blue: 0.88)
        case .white: return Color.white
        }
    }
}

enum StampPosition: String, CaseIterable, Codable, Identifiable {
    case topRight
    case topLeft
    case bottomRight
    case bottomLeft

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .topRight: return "우측 상단"
        case .topLeft: return "좌측 상단"
        case .bottomRight: return "우측 하단"
        case .bottomLeft: return "좌측 하단"
        }
    }
}

enum LetterFont: String, CaseIterable, Codable, Identifiable {
    case handwriting
    case serif
    case sans

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .handwriting: return "손글씨"
        case .serif: return "세리프"
        case .sans: return "산세리프"
        }
    }

    var font: Font {
        switch self {
        case .handwriting: return .custom("SnellRoundhand", size: 18)
        case .serif: return .serif(.body)
        case .sans: return .system(size: 18, weight: .regular, design: .default)
        }
    }
}

struct Letter: Identifiable, Codable {
    let id: UUID
    let stamp: Stamp
    let paperStyle: LetterPaperStyle
    let stampPosition: StampPosition
    let content: String
    let font: LetterFont
    let textColor: Color
    let senderName: String
    let receiverName: String
    let createdAt: Date
    let imagePath: String?

    init(
        id: UUID = UUID(),
        stamp: Stamp,
        paperStyle: LetterPaperStyle = .white,
        stampPosition: StampPosition = .topRight,
        content: String = "",
        font: LetterFont = .handwriting,
        textColor: Color = .black,
        senderName: String = "",
        receiverName: String = "",
        createdAt: Date = Date(),
        imagePath: String? = nil
    ) {
        self.id = id
        self.stamp = stamp
        self.paperStyle = paperStyle
        self.stampPosition = stampPosition
        self.content = content
        self.font = font
        self.textColor = textColor
        self.senderName = senderName
        self.receiverName = receiverName
        self.createdAt = createdAt
        self.imagePath = imagePath
    }

    var formattedDate: String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "ko_KR")
        formatter.dateFormat = "yyyy년 M월 d일"
        return formatter.string(from: createdAt)
    }

    private enum CodingKeys: String, CodingKey {
        case id
        case stamp
        case paperStyle
        case stampPosition
        case content
        case font
        case textColor
        case senderName
        case receiverName
        case createdAt
        case imagePath
        case red
        case green
        case blue
        case opacity
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(stamp, forKey: .stamp)
        try container.encode(paperStyle, forKey: .paperStyle)
        try container.encode(stampPosition, forKey: .stampPosition)
        try container.encode(content, forKey: .content)
        try container.encode(font, forKey: .font)
        try container.encode(senderName, forKey: .senderName)
        try container.encode(receiverName, forKey: .receiverName)
        try container.encode(createdAt, forKey: .createdAt)
        try container.encodeIfPresent(imagePath, forKey: .imagePath)

        let uiColor = UIColor(textColor)
        var red: CGFloat = 0
        var green: CGFloat = 0
        var blue: CGFloat = 0
        var alpha: CGFloat = 0
        uiColor.getRed(&red, green: &green, blue: &blue, alpha: &alpha)

        try container.encode(Double(red), forKey: .red)
        try container.encode(Double(green), forKey: .green)
        try container.encode(Double(blue), forKey: .blue)
        try container.encode(Double(alpha), forKey: .opacity)
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(UUID.self, forKey: .id)
        stamp = try container.decode(Stamp.self, forKey: .stamp)
        paperStyle = try container.decode(LetterPaperStyle.self, forKey: .paperStyle)
        stampPosition = try container.decode(StampPosition.self, forKey: .stampPosition)
        content = try container.decode(String.self, forKey: .content)
        font = try container.decode(LetterFont.self, forKey: .font)
        senderName = try container.decode(String.self, forKey: .senderName)
        receiverName = try container.decode(String.self, forKey: .receiverName)
        createdAt = try container.decode(Date.self, forKey: .createdAt)
        imagePath = try container.decodeIfPresent(String.self, forKey: .imagePath)

        let red = try container.decode(Double.self, forKey: .red)
        let green = try container.decode(Double.self, forKey: .green)
        let blue = try container.decode(Double.self, forKey: .blue)
        let opacity = try container.decode(Double.self, forKey: .opacity)
        textColor = Color(.sRGB, red: red, green: green, blue: blue, opacity: opacity)
    }
}
